### PR TITLE
Adding lock to node data map

### PR DESCRIPTION
**Pull request by krzysied**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Part of #81167

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -169,6 +169,41 @@ type nodeHealthData struct {
...
+func newNodeHealthMap() *nodeHealthMap {
+	return &nodeHealthMap{
+		nodeHealths: make(map[string]*nodeHealthData),
+	}
+}
+
+func (n *nodeHealthMap) get(name string) *nodeHealthData {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	return n.nodeHealths[name].deepCopy()
```
Comment (yastij) [inline pkg/controller/nodelifecycle/node_lifecycle_controller.go]: wouldn't this be simplified if we didn't return a `deepCopy()` ?

**Review by yastij — COMMENTED** [review]
/cc @wojtek-t

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 1 above)
Comment (krzysied) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: The problem is that in PR #81167 I want to add handler for pod updates. This handler and  tryUpdateNodeHealth method can potentially work on the same nodeHealthData at one time. The deep copy solves the problem of data object being updated by one function while being processed by other.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 2]
```diff
@@ -3252,19 +3252,21 @@ func TestTryUpdateNodeHealth(t *testing.T) {
...
 			_, _, currentReadyCondition, err := nodeController.tryUpdateNodeHealth(test.node)
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}
-			_, savedReadyCondition := nodeutil.GetNodeCondition(nodeController.nodeHealthMap[test.node.Name].status, v1.NodeReady)
+			_, savedReadyCondition := nodeutil.GetNodeCondition(nodeController.nodeHealthMap.get(test.node.Name).status, v1.NodeReady)
 			savedStatus := getStatus(savedReadyCondition)
 			currentStatus := getStatus(currentReadyCondition)
-			if currentStatus != savedStatus {
+			if (currentStatus == nil && savedStatus != nil) ||
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: use DeepEqual instead

File: pkg/controller/nodelifecycle/node_lifecycle_controller.go (hunk 1 above)
Comment (krzysied) [reply pkg/controller/nodelifecycle/node_lifecycle_controller.go]: Done.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go (hunk 2 above)
Comment (krzysied) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: Done.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 3]
```diff
@@ -3252,19 +3253,19 @@ func TestTryUpdateNodeHealth(t *testing.T) {
...
 			_, _, currentReadyCondition, err := nodeController.tryUpdateNodeHealth(test.node)
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}
-			_, savedReadyCondition := nodeutil.GetNodeCondition(nodeController.nodeHealthMap[test.node.Name].status, v1.NodeReady)
+			_, savedReadyCondition := nodeutil.GetNodeCondition(nodeController.nodeHealthMap.getDeepCopy(test.node.Name).status, v1.NodeReady)
 			savedStatus := getStatus(savedReadyCondition)
 			currentStatus := getStatus(currentReadyCondition)
-			if currentStatus != savedStatus {
+			if !reflect.DeepEqual(currentStatus, savedStatus) {
```
Comment (wojtek-t) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: Use api.Semantic.DeepEqual instead.

**Review by wojtek-t — COMMENTED** [review]
Just one minor comment - other than that lgtm

**Pull request merged**